httpservice/http_service.go

```diff
@@ -0,0 +1,148 @@
+}
+
+// HTTPClient contains the retrier, baseURL and the actual httpClient.
+type HTTPClient struct {
+	client      *http.Client
+	baseURL     *url.URL
+	retrier     *retry.Retrier
+	serviceName string
+}
+
+// NewHTTPClient returns a new HTTPClient for the given service name and the base url
+// The baseURL will be used to prefix the paths provided in Call and CallWithRetries
+func NewHTTPClient(serviceName string, baseURL *url.URL) *HTTPClient {
+	aggressiveTransport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   1 * time.Second, // !!!
+			KeepAlive: 90 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 3 * time.Second,
+	}
+
+	return &HTTPClient{
+		baseURL: baseURL,
+		client: &http.Client{
+			Transport: aggressiveTransport,
+		},
+		retrier: retry.NewErrorTypeRetrier(serviceName,
+			retry.DefaultBackOffOpts,
+			(*net.OpError)(nil),
+			(*RetryableHTTPError)(nil)),
+	}
+}
+
+// CallWithRetries makes calls to the network with the default retry logic
+func (client *HTTPClient) CallWithRetries(method string, path string, jsonData interface{}) ([]byte, error) {
+	res, err := client.retrier.Retry(func() (interface{}, error) {
+		return client.Call(method, path, jsonData)
+	})
+	if err != nil {
+		return []byte{}, err
+	}
+	return res.([]byte), nil
+}
+
+// Call formats the network request and returns the raw data as the response
+func (client *HTTPClient) Call(method string, path string, jsonData interface{}) ([]byte, error) {
+	req, err := client.buildJSONRequest(method, client.parseURL(path), jsonData)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 500 {
+		return nil, &RetryableHTTPError{Path: path, StatusCode: resp.StatusCode}
+	} else if resp.StatusCode >= 300 {
+		return nil, &HTTPError{Path: path, StatusCode: resp.StatusCode}
+	}
+
+	return ioutil.ReadAll(resp.Body)
+}
+
+func (client *HTTPClient) parseURL(path string) string {
+	baseURL := client.urlWithoutCreds(*client.baseURL)
+	return strings.TrimRight(baseURL, "/") + path
+}
+
+func (client *HTTPClient) urlWithoutCreds(u url.URL) string {
+	u.User = nil
+	return u.String()
+}
+
+func (client *HTTPClient) buildJSONRequest(method, path string, jsonData interface{}) (*http.Request, error) {
+
+	requestBody, err := client.marshalRequestBody(jsonData)
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := http.NewRequest(method, path, requestBody)
+	if err != nil {
+		return nil, err
+	}
+
+	if requestBody != nil {
+		request.Header.Set("content-type", "application/json")
+	}
+
+	return request, nil
+}
+
+func (client *HTTPClient) marshalRequestBody(jsonData interface{}) (io.Reader, error) {
+	if jsonData == nil {
+		return nil, nil
+	}
+	body, err := json.Marshal(jsonData)
+	if err != nil {
+		return nil, err
+	}
+	return strings.NewReader(string(body)), nil
```

Comment:
It'd be better to use `bytes.NewReader(body)` here. `string(body)` actually does some allocations and has performance implications if I remember correctly.